exp/runtime/internal/controllers/extension_controller.go

```diff
@@ -21,74 +21,169 @@ import (
+		Named("runtime/extension").
 		WithOptions(options).
+		WithEventFilter(predicates.ResourceNotPausedAndHasFilterLabel(ctrl.LoggerFrom(ctx), r.WatchFilterValue)).
 		Complete(r)
-
 	if err != nil {
 		return errors.Wrap(err, "failed setting up with a controller manager")
 	}
 	return nil
 }
 
-func (r *ExtensionReconciler) Reconcile(ctx context.Context, req ctrl.Request) (_ ctrl.Result, reterr error) {
+func (r *Reconciler) Reconcile(ctx context.Context, req ctrl.Request) (_ ctrl.Result, reterr error) {
 	log := ctrl.LoggerFrom(ctx)
-	log.V(4).Info("Reconciling")
+	if !r.RuntimeClient.IsRegistryReady() {
+		err := r.syncRegistry(ctx)
+		if err != nil {
+			return ctrl.Result{}, errors.Wrapf(err, "extensions controller not initialized")
+		}
+		// Note: there was an intermittent bug when allowing this to continue (likely due to patching the same object multiple times?)
+		return ctrl.Result{}, nil
+	}
 
-	ext := &expv1.Extension{}
-	err := r.Client.Get(ctx, req.NamespacedName, ext)
+	log.V(4).Info("Reconciling")
+	extension := &runtimev1.Extension{}
+	err := r.Client.Get(ctx, req.NamespacedName, extension)
 	if err != nil {
 		if apierrors.IsNotFound(err) {
-			// Object not found, return.  Created objects are automatically garbage collected.
-			// For additional cleanup logic use finalizers.
+			// Object not found, return.
 			return ctrl.Result{}, nil
 		}
 
 		// Error reading the object - requeue the request.
 		return ctrl.Result{}, err
 	}
 
+	// Initialize the patch helper
+	patchHelper, err := patch.NewHelper(extension, r.Client)
+	if err != nil {
+		return ctrl.Result{}, err
+	}
+
+	// FIXME: could dedupe these patch calls. Ignoring for simplicity at this stage.
+	defer func() {
+		// Always attempt to patch the object and status after each reconciliation.
+		// Patch ObservedGeneration only if the reconciliation completed successfully
+		patchOpts := []patch.Option{}
+		if reterr == nil {
+			patchOpts = append(patchOpts, patch.WithStatusObservedGeneration{},
+				patch.WithOwnedConditions{Conditions: []clusterv1.ConditionType{
+					runtimev1.RuntimeExtensionDiscovered,
+				}})
+		}
+		if err := patchExtension(ctx, patchHelper, extension, patchOpts...); err != nil {
+			reterr = kerrors.NewAggregate([]error{reterr, err})
+		}
+	}()
+
 	// Handle deletion reconciliation loop.
-	if !ext.ObjectMeta.DeletionTimestamp.IsZero() {
-		return r.reconcileDelete(ctx, ext)
+	if !extension.ObjectMeta.DeletionTimestamp.IsZero() {
+		return r.reconcileDelete(ctx, extension)
 	}
 
 	// Handle normal reconciliation loop.
-	return r.reconcile(ctx, ext)
+	return r.reconcile(ctx, extension)
 }
 
-func (r *ExtensionReconciler) reconcileDelete(ctx context.Context, ext *expv1.Extension) (ctrl.Result, error) {
-	r.Registry.RemoveRuntimeExtension(ext)
+func patchExtension(ctx context.Context, patchHelper *patch.Helper, ext *runtimev1.Extension, options ...patch.Option) error {
+	// Can add errors and waits here to the Extension object.
```

Comment:
Can you elaborate a bit? I'm not sure I get what we would add here

Reply:
This is about possibly adding conditions to the extension object - out of scope for now.